pm: Make device pm optional per state - V2

ceolin opened the pull request:
This is a proposal to make device power management (triggered when SoC is idle) completely optional.

- Targets can select which power states should ***NOT***  trigger device power management
- If there is no power state with this property enabled, the pm subsystem won't trigger device power management
- Removed the need for `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`

Notes:

This is an alternative version to https://github.com/zephyrproject-rtos/zephyr/pull/60939. This version is less intrusive and keep the current behavior. When device runtime pm is enabled, system managed device power management is disabled (can be enabled though).

In this version device pm is triggered by default and can have it disabled individually (instead of by default does not trigger device power management)

Additional context:
https://github.com/zephyrproject-rtos/zephyr/pull/60463
https://github.com/zephyrproject-rtos/zephyr/pull/60369
https://github.com/zephyrproject-rtos/zephyr/pull/60939

Commit:
pm: Deprecate PM_DEVICE_RUNTIME_EXCLUSIVE

That is option has shown confusing on it is attempt to prevent
system pm doing device power management. Lets address this
problem properly.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Disable device pm per power state

Make it possible to disble device power management individually per
power state.  This allows targets tuning which states should
(and which should not) trigger device power management.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: Exclude device pm path when it is not needed

Remove device pm path when there is no is no power state in DT with
device pm enabled. This basically does the same thing that was done
by PM_DEVICE_RUNTIME_EXCLUSIVE.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: power_mgmt: Use DTS overlay for power state

Define the power state needed in the test in DT.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: device_wakeup_api: Use DTS overlay for power state

Define the power state needed in the test in DT.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
doc: pm: Update system device pm

Add new information about new property to
enable/disable device power management per
power state.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
pm: device: De-couple device pm from system pm

PM_DEVICE is not attached to system managed device power management.
It is a very common use case targets with device runtime power
management that don't want system device power management enabled.

We introduce a new symbol (PM_DEVICE_SYSTEM_MANAGED) to explicit
control whether or not system device power management should be
globally enabled.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: power_mgmt: Add test for PM_DEVICE_SYSTEM_MANAGED

Add a new test that disables PM_DEVICE_SYSTEM_MANAGED, since this
option is enabled by default.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

Commit:
tests: pm: Replace depracated build option

Replace the usage of CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE with
CONFIG_PM_DEVICE_SYSTEM_MANAGED.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
@JordanYates @erwango @nashif can you take a look on this please ?

ceolin commented:
folks, can I get some feedback here ? the original changes was already discussed in appropriated forum and there were no push-backs. This version is simpler and kept the current behavior but it adds more flexibility and organize device pm imho.

If there is no major block I would like to have it sooner than later to avoid regressions close to the release.

JordanYates commented:
> @JordanYates @erwango @nashif can you take a look on this please ?

I have no experience working with system PM, I don't feel I can provide any useful feedback.

ceolin commented:
> > @JordanYates @erwango @nashif can you take a look on this please ?
> 
> I have no experience working with system PM, I don't feel I can provide any useful feedback.

Sure thing. Thanks for replying it :)

erwango commented on dts/bindings/power/zephyr,power-state.yaml:
```diff
@@ -31,3 +31,9 @@ properties:
+  zephyr,pm-device-disabled:
+    type: boolean
+    description: |
+       Disable system managed device power management for this state. When this
+       is enabled the power management subsystem will not suspend devices before entering
```
```suggestion
       Disable system managed device power management for this state. When set,
       the power management subsystem will not suspend devices before entering
```

erwango commented on include/zephyr/pm/state.h:
```diff
@@ -209,6 +219,7 @@ struct pm_state_info {
 		.substate_id = DT_PROP_OR(node_id, substate_id, 0),	       \
 		.min_residency_us = DT_PROP_OR(node_id, min_residency_us, 0),  \
 		.exit_latency_us = DT_PROP_OR(node_id, exit_latency_us, 0),    \
+		.pm_device_disabled = DT_PROP_OR(node_id, zephyr_pm_device_disabled, true),    \
```
the _OR will set `pm_device_disabled` to `true` when `zephyr,pm-device-disabled` is not set.
Since `zephyr,pm-device-disabled` is a boolean, you can use:
```suggestion
		.pm_device_disabled = DT_PROP(node_id, zephyr_pm_device_disabled),    \
```
which will set `pm_device_disabled` to true when `zephyr,pm-device-disabled` is present and vice versa.

erwango reviewed (dismissed):
Please rebase.

Some addiotinal comments:
You're mentioning https://github.com/zephyrproject-rtos/zephyr/pull/60369, but I don't see a direct link, can you elaborate on this point ?
It would be nice to have example of use case where it is interesting to make use of System Managed Device Power Management in some states and not on others.
(maybe both are linked, but I've not been able to connect the dots).

ceolin commented on subsys/pm/device_system_managed.c:
```diff
@@ -0,0 +1,90 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(pm_device, CONFIG_PM_DEVICE_LOG_LEVEL);
+
+#define DT_PM_DEVICE_ENABLED(node_id)						\
```
Nope, because what I am checking here is that we don't have *all* states disabling device power management.

Commit:
tests: power_mgmt: Test system-managed disabled per state

Check that a power state that has system-managed device power
management disabled does not trigger device power management
when the system sleeps.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

ceolin commented:
> Please rebase.
> 
> Some addiotinal comments: You're mentioning #60369, but I don't see a direct link, can you elaborate on this point ? It would be nice to have example of use case where it is interesting to make use of System Managed Device Power 

At some point the author mentioned that some power states shouldn't trigger device power management. I don't remember exactly the full context though.

What this pr does is basically organize device power management and make clear that there are two methods, system-managed and runtime.  It keeps exactly the same current behavior (if it doesn't it is a bug and I can fix it). The additional feature is just providing a way to fine tuning  system-managed device pm allowing one to disable this method per power state. We could do something like:
```
/ {
	cpus {
		power-states {
			state0: state0 {
				compatible = "zephyr,power-state";
				power-state-name = "suspend-to-idle";
			};
			state1: state1 {
				compatible = "zephyr,power-state";
				power-state-name = "standby";
				zephyr,pm-device-disabled;
			};
			state2: state2 {
				compatible = "zephyr,power-state";
				power-state-name = "standby";
			};
		};
	};
};

&cpu1 {
	cpu-power-states = <&state0 &state1>;
};
&cpu0 {
	cpu-power-states = <&state0 &state2>;
};
```

In this example only in the core devices will be suspended when the cpu transition to `standby`.
That is just to give an idea, the idea is that one can just disable device power management in any state that makes sense to their target.

ceolin commented:
@erwango ping :)

ceolin commented:
@erwango done. Can you take a look again please ?

erwango commented on subsys/pm/Kconfig:
```diff
@@ -78,12 +78,11 @@ config PM_DEVICE
+	  management subsystem. This allows device drivers to do any
+	  necessary power management operations like turning off
+	  device clocks and peripherals. Device drivers may also save
+	  and restore states in these hook functions.
```
Non blocking: We could use more detailed information on `states` here.

ceolin replied:
that is hard to be specific without the risk of saying something that can be incomplete or wrong for a particular device :/

erwango commented on subsys/pm/device_system_managed.c:
```diff
@@ -0,0 +1,89 @@
+LOG_MODULE_DECLARE(pm_device, CONFIG_PM_DEVICE_LOG_LEVEL);
+
+#define DT_PM_DEVICE_ENABLED(node_id)						\
+	COND_CODE_1(DT_PROP_OR(node_id, zephyr_pm_device_disabled, 0),	\
```
Non blocking:
```suggestion
	COND_CODE_1(DT_PROP(node_id, zephyr_pm_device_disabled),	\
```
should be used when prop is a boolean

erwango commented on doc/services/pm/device.rst:
```diff
@@ -72,16 +72,46 @@ For more information, see :ref:`pm-device-runtime`.
 from a sleep state, devices are resumed in the opposite order that they were
 suspended.
 
+The decision about suspending devices in a low power state is done based on the
+state and if it has setting the property ``zephyr,pm-device-disabled``. Here is
```
```suggestion
The decision about suspending devices when entering a low power state is done based on the
state and if it has set the property ``zephyr,pm-device-disabled``. Here is
```

erwango commented on doc/services/pm/device.rst:
```diff
@@ -72,16 +72,46 @@ For more information, see :ref:`pm-device-runtime`.
 
+The decision about suspending devices in a low power state is done based on the
+state and if it has setting the property ``zephyr,pm-device-disabled``. Here is
+an example of a target with two low power states with only triggering device power
+management:
```
```suggestion
an example of a target with two low power states with only one triggering device power
management:
```

ceolin commented:
@erwango thanks a lot for reviewing it. Have updated it with the suggestions. Would you mind look it again ?

mmahadevan108 commented on subsys/pm/device_system_managed.h:
```diff
@@ -0,0 +1,22 @@
+#else
+
+bool pm_resume_devices(void) { return true; }
+void pm_suspend_devices(void) {}
```
@ceolin, is this a bug? Why are the function definitions different in the #if and #else cases?
